### ISSUE TITLE
Literal math failing back for simple case

### ISSUE THREAD

**Issue by JanEricNitschke** [microsoft/pyright#5020]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Literal math is falling back to `LiteralString` for a very simple case.

**To Reproduce**
This is the long example to reproduce what i am getting with the [pyright playground](https://pyright-playground.decorator-factory.su/?gzip=H4sIAFJ2SmQC_6WUTU-EMBCG7_0VE7xArBtoi4cmGi_GiwdjjBezITVblMhSHMBkY_zvlq9FiiYbt4fJAPM-0zIvpGi2UO_KrHiBbFsarOE2qzWqnMLDrtQ3jcINBfOhMTdqQ8hGp5CUaEqNyVuW55XfRQlZUQdwdjmpngbQU0ghosAocAqCQrxeSwJ2oa4bLKDTWzmgKl60fx4QQq72Dbt-lp3UJqlqtPuct903WXfdx0sv9IYuq9XqP7zI4UUz3j-AzAGyY4HcAfJjgcIBimOBsQOM58A_MV2NO1_XRMSdNwU7JBtYG3gbRBv2TbN0sNrFBYT9rR8utIBFVbSsipZVTAKcQGHelYT7Jk39u9t7FoYiWIjZUswPFvOlWBwsFktxfLA47sWoskrDo8obfY1o0Pd61quqoDbwrKdvOKLngdd-x2iaYpMMPeEzkhFlklEuORVS0FjGX4SkBqHM1U4jhSKZfgeTdmWnvK38YJqHPYm_v2jX_J80YKazBN1pXRPNCJ1iMKLd7O_WXIDblZvx-VA82bZ_R9aXfcLGhI-JGBPrU9u1zz04nW3nG8U9u9imBQAA):
```
from typing import Literal, TypeGuard, overload

def _proper_kills(kills: int) -> TypeGuard[Literal[0, 1, 2, 3, 4, 5]]:
    return kills in range(6)


@overload
def _int_to_string_kills(kills: Literal[0]) -> Literal["0"]:
    ...

@overload
def _int_to_string_kills(kills: Literal[1]) -> Literal["1"]:
    ...


@overload
def _int_to_string_kills(kills: Literal[2]) -> Literal["2"]:
    ...


@overload
def _int_to_string_kills(kills: Literal[3]) -> Literal["3"]:
    ...


@overload
def _int_to_string_kills(kills: Literal[4]) -> Literal["4"]:
    ...


@overload
def _int_to_string_kills(kills: Literal[5]) -> Literal["5"]:
    ...


def _int_to_string_kills(
    kills: Literal[0, 1, 2, 3, 4, 5]
) -> Literal["0", "1", "2", "3", "4", "5"]:
    if kills == 0:
        return "0"
    if kills == 1:
        return "1"
    if kills == 2:  # noqa: Ruff(PLR2004)
        return "2"
    if kills == 3:  # noqa: Ruff(PLR2004)
        return "3"
    if kills == 4:  # noqa: Ruff(PLR2004)
        return "4"
    if kills == 5:  # noqa: Ruff(PLR2004)
        return "5"
    raise ValueError("kills has to be in range(1,6)")


round_kills = {1:1,2:2,3:3,4:4,5:5}

for player, n_kills in round_kills.items():
        if (
            _proper_kills(n_kills)
        ):  # 0, 1, 2, 3, 4, 5
            kills_string =  _int_to_string_kills(n_kills)
            long_kills_string: Literal["kills1","kills2","kills3","kills4","kills5"] = "kills" + kills_string
```
**Expected behavior**
The expected behavior is that pyright infers target to be valid for type Literal["kills1","kills2","kills3","kills4","kills5"] but it infers `LiteralString` and complains about the line.

For me locally it even happens if i directly set:

`kills_string: Literal["0","1","2","3","4","5"]` but that is not the case on the playground.

**VS Code extension or command-line**
Happens for pylance vscode extension v2023.4.40 as well as with the CI: pyright==1.1.305

**Comment by erictraut** [comment]
This is as designed. Literal math is disabled in a loop because it is difficult to impossible to determine whether the value is modified each time in the loop. For example:

```python
x = 1
reveal_type(x) # Literal[1]
x += 1
reveal_type(x) # Literal[2]
for n in range(10):
    x += 1
    reveal_type(x) # int
```

**Comment by JanEricNitschke** [comment]
Seems a bit weird that that is also the case even if i apply the TypeGuard within the loop and do the literal math as the next operation.

But i guess the work around is:

```
round_kills = {1:1,2:2,3:3,4:4,5:5}

for player, n_kills in round_kills.items():
        if (
            _proper_kills(n_kills)
        ):  # 0, 1, 2, 3, 4, 5
            kills_string =  _int_to_string_kills(n_kills)
            long_kills_string: Literal["kills1","kills2","kills3","kills4","kills5"] = "kills" + kills_string

do_stuff(n_kills: int) -> None:
     kills_string =  _int_to_string_kills(n_kills)
     long_kills_string: Literal["kills1","kills2","kills3","kills4","kills5"] = "kills" + kills_string
```
